File: midonet-api/src/main/java/org/midonet/api/auth/keystone/v2_0/KeystoneClient.java
/*
 * Copyright 2012 Midokura PTE LTD.
 */
package org.midonet.api.auth.keystone.v2_0;

import com.google.inject.Inject;
import com.sun.jersey.api.client.Client;
import com.sun.jersey.api.client.ClientHandlerException;
import com.sun.jersey.api.client.UniformInterfaceException;
import com.sun.jersey.api.client.WebResource;
import org.midonet.api.auth.keystone.KeystoneBadCredsException;
import org.midonet.api.auth.keystone.KeystoneConnectionException;
import org.midonet.api.auth.keystone.KeystoneServerException;
import org.midonet.util.StringUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

/**
 * Keystone Client V2.0
 */
public class KeystoneClient {

    private final static Logger log = LoggerFactory
            .getLogger(KeystoneClient.class);

    private final int port;
    private final String protocol;
    private final String host;
    private final String adminToken;

    public static final String KEYSTONE_TOKEN_HEADER_KEY = "X-Auth-Token";
    public static final String MARKER_QUERY = "marker";
    public static final String LIMIT_QUERY = "limit";

    @Inject
    public KeystoneClient(String host, int port, String protocol,
                          String adminToken) {
        this.host = host;
        this.port = port;
        this.protocol = protocol;
        this.adminToken = adminToken;
    }

    /**
     * @return the serviceUrl
     */
    public String getServiceUrl() {
        return new StringBuilder(this.protocol).append("://")
                .append(this.host).append(":").append(
                        Integer.toString(this.port))
                .append("/v2.0").toString();
    }

    /**
     * @return the tokens URL
     */
    private String getTokensUrl() {
        return new StringBuilder(getServiceUrl()).append("/tokens").toString();
    }

    /**
     * @return the tokens URL
     *
     * @param token Token to append to the URL
     */
    private String getTokensUrl(String token) {
        return new StringBuilder(getTokensUrl()).append("/").append(token)
                .toString();
    }

    private String getTenantsUrl() {
        return new StringBuilder(getServiceUrl()).append("/tenants").toString();
    }

    private String getTenantUrl(String id) {
        return new StringBuilder(getTenantsUrl()).append("/").append(
                id).toString();
    }

    private <T> T sendGetRequest(WebResource resource, Class<T> clazz)
            throws KeystoneServerException, KeystoneConnectionException {

        try {
            return resource.accept(MediaType.APPLICATION_JSON)
                    .header(KEYSTONE_TOKEN_HEADER_KEY, this.adminToken)
                    .get(clazz);
        } catch (UniformInterfaceException e) {
            throw new KeystoneServerException("Keystone server error.", e,
                    e.getResponse().getStatus());
        } catch (ClientHandlerException e) {
            throw new KeystoneConnectionException(
                    "Could not connect to Keystone server. Uri="
                            + resource.getURI(), e);
        }
    }


    public KeystoneAccess createToken(KeystoneAuthCredentials credentials)
            throws KeystoneServerException, KeystoneConnectionException,
            KeystoneBadCredsException {

        Client client = Client.create();
        WebResource resource = client.resource(getTokensUrl());

        try {
            return resource.type(MediaType.APPLICATION_JSON)
                    .accept(MediaType.APPLICATION_JSON)
                    .post(KeystoneAccess.class, credentials);
        } catch (UniformInterfaceException e) {
            if (e.getResponse().getStatus()
                == Response.Status.BAD_REQUEST.getStatusCode()) {
                String err = "KeystoneClient: keystone login creds invalid "
                              + credentials.toString();
                log.warn(err);
                throw new KeystoneBadCredsException(err, e);
            } else if (e.getResponse().getStatus()
                == Response.Status.UNAUTHORIZED.getStatusCode()) {
                String err = "KeystoneClient: keystone login creds not found "
                              + credentials.toString();
                log.warn(err,credentials);
                throw new KeystoneBadCredsException(err, e);
            } else {
                throw new KeystoneServerException("Keystone server error.", e,
                    e.getResponse().getStatus());
            }
        } catch (ClientHandlerException e) {
            throw new KeystoneConnectionException(
                    "Could not connect to Keystone server. Uri="
                            + resource.getURI(), e);
        }
    }

    public KeystoneAccess getToken(String token) throws KeystoneServerException,
            KeystoneConnectionException {

        Client client = Client.create();
        WebResource resource = client.resource(getTokensUrl(token));

        try {
            return sendGetRequest(resource, KeystoneAccess.class);
        } catch (KeystoneServerException ex) {
            if (ex.getStatus() == Response.Status.NOT_FOUND
                    .getStatusCode()) {
                // This indicates that the token was not found
                log.warn("KeystoneClient: Token not found {}", token);
                return null;
            }

            throw ex;
        }
    }

    /**
     * Retrieves from Keystone server a tenant given its ID.
     *
     * @param id  ID of the tenant
     * @return {@link KeystoneTenant} object
     * @throws KeystoneServerException
     * @throws KeystoneConnectionException
     */
    public KeystoneTenant getTenant(String id) throws KeystoneServerException,
            KeystoneConnectionException {
        log.debug("KeystoneClient.getTenant: entered id=" + id);

        Client client = Client.create();
        WebResource resource = client.resource(getTenantUrl(id));

        try {
            return sendGetRequest(resource, KeystoneTenant.class);
        } catch (KeystoneServerException ex) {
            if (ex.getStatus() == Response.Status.NOT_FOUND.getStatusCode()) {
                // This indicates that the tenant was not found.  Not an error
                // from MidoNet.
                log.info("KeystoneClient: Tenant not found {}", id);
                return null;
            }

            throw ex;
        }
    }

    /**
     * Retrieves from Keystone server a list of tenants from the given optional
     * marker and limit values.  The marker and limit values are appended to
     * the actual HTTP request to Keystone as query string parameters.  The
     * return object of {@link KeystoneTenantList} represents the response body
     * from the Keystone API containing the tenant list.
     *
     * @param marker The ID of the tenant fetched in the previous request.
     *               The returned list will start from the tenant after the
     *               tenant with this ID.  A null and empty values are ignored.
     * @param limit  The number of tenants to fetch.  Must be greater than 0.
     *               This field is ignored if the value is less than 1.
     * @return {@link KeystoneTenantList} object, representing the response from
     *         the Keystone API.
     * @throws KeystoneServerException
     * @throws KeystoneConnectionException
     */
    public KeystoneTenantList getTenants(String marker, int limit)
            throws KeystoneServerException, KeystoneConnectionException {
        log.debug("KeystoneClient.getTenants: entered marker=" + marker +
                ", limit=" + limit);

        Client client = Client.create();
        WebResource resource = client.resource(getTenantsUrl());

        if (!StringUtil.isNullOrEmpty(marker)) {
            resource = resource.queryParam(MARKER_QUERY, marker);
        }

        if (limit > 0) {
            resource = resource.queryParam(LIMIT_QUERY,
                    Integer.toString(limit));
        }

        return sendGetRequest(resource, KeystoneTenantList.class);
    }
}

File: midonet-cluster/src/main/scala/org/midonet/cluster/auth/keystone/KeystoneClient.scala
/*
 * Copyright 2016 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.auth.keystone

import java.net.URI
import java.time.format.DateTimeFormatter
import java.time.{LocalDateTime, ZoneOffset, ZonedDateTime}
import java.util.Collections
import java.util.logging.Logger

import javax.annotation.Nullable
import javax.ws.rs.core.MediaType.APPLICATION_JSON
import javax.ws.rs.core.Response.Status
import javax.ws.rs.core.UriBuilder

import scala.annotation.tailrec
import scala.concurrent.duration._
import scala.util.Try

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.jaxrs.json.JacksonJaxbJsonProvider
import com.sun.jersey.api.client.config.DefaultClientConfig
import com.sun.jersey.api.client.filter.LoggingFilter
import com.sun.jersey.api.client.{Client, ClientHandlerException, ClientResponse, UniformInterfaceException}

import org.apache.commons.lang.StringUtils
import org.apache.commons.lang3.StringUtils.{isBlank, isNotBlank}

import org.midonet.cluster.auth.keystone.KeystoneClient.{AdminToken, TokenExpirationGuard, parseExpiresAt, parseTimestamp}

object KeystoneClient {

    /**
      * This ensures that clock difference between the client and the Keystone
      * server does not cause the client to use an expired token.
      */
    private val TokenExpirationGuard = 1 minute

    /**
      * Contains information about an administrative token.
      */
    private case class AdminToken(id: String,
                                  serverIssueTime: Long,
                                  clientIssueTime: Long,
                                  expirationTime: Long,
                                  isStatic: Boolean)

    /**
      * Parses the given timestamp from the ISO8601 date format to a Unix
      * timestamp. If any parsing fails, the method throws a [[KeystoneException]].
      */
    @throws[KeystoneException]
    private[keystone] def parseTimestamp(string: String): Long = {
        val tryZonedTime = Try {
            ZonedDateTime.parse(string, DateTimeFormatter.ISO_DATE_TIME)
                         .toInstant.toEpochMilli
        }
        val tryLocalTime = Try {
            LocalDateTime.parse(string, DateTimeFormatter.ISO_DATE_TIME)
                         .toInstant(ZoneOffset.of("Z")).toEpochMilli
        }

        tryZonedTime recoverWith {
            case _ => tryLocalTime
        } getOrElse {
            throw new KeystoneException(
                null, s"Unknown token timestamp format for $string", null)
        }
    }

    /**
      * Parses the given expiration timestamp string from the ISO8601 date
      * format to a Unix timestamp. If the string is `null`, meaning the
      * token never expires, the method returns the maximum long value.
      */
    @throws[KeystoneException]
    private[keystone] def parseExpiresAt(@Nullable string: String): Long = {
        if (string eq null)
            Long.MaxValue
        else
            parseTimestamp(string)
    }
}


class KeystoneClient(config: KeystoneConfig) {

    private val jsonProvider = new JacksonJaxbJsonProvider()
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

    private val clientConfig = new DefaultClientConfig()
    clientConfig.getSingletons.add(jsonProvider)

    private val client = Client.create(clientConfig)
    client.addFilter(new LoggingFilter(
        Logger.getLogger("org.midonet.cluster.auth.keystone")))

    @volatile private var adminToken =
        if (isNotBlank(config.adminToken))
            AdminToken(config.adminToken, 0L, 0L, Long.MaxValue, isStatic = true)
        else null

    /**
      * Gets the current Keystone version.
      */
    @throws[KeystoneException]
    def getVersion: KeystoneVersion = {
        get(classOf[KeystoneVersion], None, None)()
    }

    /**
      * Authenticates against Keystone for the given tenant using the specified
      * password credentials. The method returns an [[AuthResponse]] instance
      * containing an authorization token for the given user.
      */
    @throws[KeystoneException]
    def authenticate(tenantName: String, userName: String, password: String)
    : AuthResponse = withVersion {
        case 2 =>
            val auth = v2.KeystoneAuth(v2.PasswordAuth(
                v2.PasswordCredentials(userName, password),tenantName))
            val access = post(auth, classOf[v2.KeystoneAccess], token = None,
                              "tokens")()
            AuthResponse(tokenId = access.access.token.id, r2 = access.access)
        case 3 =>
            val domain =
                if (StringUtils.isNotBlank(config.domainId))
                    v3.Domain(id = config.domainId)
                else
                    v3.Domain(name = config.domainName)
            val auth = v3.KeystoneAuth(v3.Auth(
                v3.Identity(methods = Collections.singletonList("password"),
                            password = v3.Password(v3.User(
                                name = userName,
                                password = password,
                                domain = domain))),
                v3.Scope(v3.Project(name = tenantName,
                                    domain = domain))))
            handleAuthResponse(post(auth, classOf[ClientResponse], token = None,
                                    "auth", "tokens")())
    }

    /**
      * Authenticates against Keystone for the given tenant using the specified
      * token. The method returns an [[AuthResponse]] instance containing an
      * authorization token for the user of the given token.
      */
    @throws[KeystoneException]
    def authenticate(tenantName: String, token: String): AuthResponse = withVersion {
        case 2 =>
            val access = post(v2.KeystoneAuth(v2.TokenAuth(v2.Token(id = token),
                                                           tenantName)),
                              classOf[v2.KeystoneAccess], token = None, "tokens")()
            AuthResponse(tokenId = access.access.token.id, r2 = access.access)
        case 3 =>
            val auth = v3.KeystoneAuth(v3.Auth(v3.Identity(
                methods = Collections.singletonList("token"),
                token = v3.Token(id = token))))
            handleAuthResponse(post(auth, classOf[ClientResponse], token = None,
                                    "auth", "tokens")())
    }

    /**
      * Validates an authorization token using the [[KeystoneClient]]'s
      * administrative credentials. If the `tenantScope` argument is true, then
      * the method validates the token in the specified tenant scope for
      * performance.
      */
    @throws[KeystoneException]
    def validate(token: String, tenantScope: Option[String] = None): AuthResponse = {
        var params = Seq.empty[(String, String)]
        if (tenantScope.nonEmpty)
            params = params :+ ("belongsTo", tenantScope.get)
        withAdminToken() { adminToken =>
            withVersion {
                case 2 =>
                    val access = get(classOf[v2.KeystoneAccess], Some(adminToken),
                                     None, "tokens", token)(params: _*)
                    AuthResponse(tokenId = access.access.token.id,
                                 r2 = access.access)
                case 3 =>
                    handleAuthResponse(get(classOf[ClientResponse], Some(adminToken),
                                           Some(token), "auth", "tokens")())
            }
        }
    }

    /**
      * Lists the users for the given client token. The client must be an
      * administrator.
      */
    @throws[KeystoneException]
    def listUsers(token: String): v2.KeystoneUsers = {
        withVersion {
            case 2 =>
                get(classOf[v2.KeystoneUsers], Some(token), subjectToken = None,
                    "users")()
        }
    }

    /**
      * Gets the information for the user with the specified name. The client
      * must be an administrator.
      */
    @throws[KeystoneException]
    def getUserByName(name: String, token: String): v2.KeystoneUser = {
        withVersion {
            case 2 =>
                get(classOf[v2.KeystoneUser], Some(token), subjectToken = None,
                    "users")(("name", name))
        }
    }

    /**
      * Gets the information for the user with the specified identifier. The
      * client must be an administrator.
      */
    @throws[KeystoneException]
    def getUserById(id: String, token: String): v2.KeystoneUser = {
        withVersion {
            case 2 =>
                get(classOf[v2.KeystoneUser], Some(token), subjectToken = None,
                    "users", id)()
        }
    }

    /**
      * Lists the tenants for the given client token. The client must be an
      * administrator.
      * @param marker Specifies the identifier of the last seen tenant when using
      *               multi-page listing.
      * @param limit Specifies a limit on the number of items returned by
      *              the operation when using multi-page listing.
      */
    @throws[KeystoneException]
    def listTenants(token: String, marker: Option[String] = None,
                    limit: Option[Int] = None)
    : v2.KeystoneTenants = {
        withVersion {
            case 2 =>
                var params = Seq.empty[(String, String)]
                if (marker.isDefined) params = params :+("marker", marker.get)
                if (limit.isDefined) params = params :+("limit", limit.get.toString)
                get(classOf[v2.KeystoneTenants], Some(token), subjectToken = None,
                    "tenants")(params: _*)
        }
    }

    /**
      * Gets the information for the tenant with the specified name. The client
      * must be an administrator.
      */
    @throws[KeystoneException]
    def getTenantByName(name: String, token: String)
    : v2.KeystoneTenant = {
        withVersion {
            case 2 =>
                get(classOf[v2.KeystoneTenant], Some(token), subjectToken = None,
                    "tenants")(("name", name))
        }
    }

    /**
      * Gets the information for the tenant with the specified identifier. The
      * client must be an administrator.
      */
    @throws[KeystoneException]
    def getTenantById(id: String, token: String)
    : v2.KeystoneTenant = {
        withVersion {
            case 2 =>
                get(classOf[v2.KeystoneTenant], Some(token), subjectToken = None,
                    "tenants", id)()
        }
    }

    /**
      * Gets the list of roles for the user on the specified tenant. The client
      * must be an administrator.
      */
    @throws[KeystoneException]
    def getTenantUserRoles(tenantId: String, userId: String, token: String,
                           marker: Option[String] = None,
                           limit: Option[Int] = None)
    : v2.KeystoneRoles = withVersion {
        case 2 =>
            var params = Seq.empty[(String, String)]
            if (marker.isDefined) params = params :+ ("marker", marker.get)
            if (limit.isDefined) params = params :+ ("limit", limit.get.toString)
            get(classOf[v2.KeystoneRoles], Some(token), subjectToken = None,
                "tenants", tenantId, "users", userId, "roles")(params: _*)
    }

    /**
      * Lists the projects for the given client token. The client must be an
      * administrator.
      */
    @throws[KeystoneException]
    def listProjects(token: String): v3.KeystoneProjects = {
        withVersion {
            case 3 =>
                get(classOf[v3.KeystoneProjects], Some(token), subjectToken = None,
                    "projects")()
        }
    }

    /**
      * Gets the information for the project with the specified name. The client
      * must be an administrator.
      */
    @throws[KeystoneException]
    def getProjectByName(name: String, token: String): v3.KeystoneProjects = {
        withVersion {
            case 3 =>
                get(classOf[v3.KeystoneProjects], Some(token), subjectToken = None,
                    "projects")(("name", name))
        }
    }

    /**
      * Gets the information for the project with the specified identifier. The
      * client must be an administrator.
      */
    @throws[KeystoneException]
    def getProjectById(id: String, token: String): v3.KeystoneProject = {
        withVersion {
            case 3 =>
                get(classOf[v3.KeystoneProject], Some(token), subjectToken = None,
                    "projects", id)()
        }
    }

    /**
      * Gets the Keystone URI for the given path segments and query
      * parameters.
      */
    private def uriFor(paths: String*)(params: (String, String)*): URI = {
        val version = withVersion { case 2 => "v2.0" case 3 => "v3" }
        val uriPath =
            if (StringUtils.isNotBlank(config.urlOverride))
                config.urlOverride
            else
                s"${config.protocol}://${config.host}:${config.port}/$version"

        val builder = UriBuilder
            .fromPath(uriPath)
            .segment(paths: _*)
        for ((param, value) <- params)
            builder.queryParam(param, value)
        builder.build()
    }

    @throws[KeystoneException]
    @inline
    private def withVersion[R](pf: PartialFunction[Int, R]): R = {
        pf.applyOrElse(
            config.version,
            (v: Int) =>
                throw new KeystoneException(
                    null, s"Unsupported Keystone version ${config.version}", null))
    }

    /**
      * A GET HTTP request to the Keystone server.
      */
    private def get[T](clazz: Class[T], authToken: Option[String],
                       subjectToken: Option[String], paths: String*)
                      (params: (String, String)*): T = {
        val uri = uriFor(paths: _*)(params: _*)
        tryRequest(uri) {
            val request = client.resource(uri)
                                .`type`(APPLICATION_JSON)
                                .accept(APPLICATION_JSON)
            if (authToken.isDefined) {
                request.header("X-Auth-Token", authToken.get)
            }
            if (subjectToken.isDefined) {
                request.header("X-Subject-Token", subjectToken.get)
            }
            request.get(clazz)
        }
    }

    /**
      * A POST HTTP request to the Keystone server.
      */
    private def post[T, R](obj: T, clazz: Class[R], token: Option[String],
                           paths: String*)
                          (params: (String, String)*): R = {
        val uri = uriFor(paths: _*)(params: _*)
        tryRequest(uri) {
            val request = client.resource(uri)
                                .`type`(APPLICATION_JSON)
                                .accept(APPLICATION_JSON)
            if (token.isDefined) {
                request.header("X-Auth-Token", token)
            }
            request.post(clazz, obj)
        }
    }

    /**
      * Provides exception handling for the given function.
      */
    private def tryRequest[R](uri: URI)(f: => R): R = {
        val uriString = if (uri ne null) uri.toString else null
        try {
            f
        } catch {
            case e: UniformInterfaceException
                if e.getResponse.getStatus == Status.UNAUTHORIZED.getStatusCode =>
                throw new KeystoneUnauthorizedException(uriString, e)
            case e: UniformInterfaceException =>
                throw new KeystoneException(uriString, e)
            case e: ClientHandlerException =>
                throw new KeystoneConnectionException(uriString, e)
        }
    }

    /**
      * Calls a function with a current administrative token.
      */
    @tailrec
    private def withAdminToken[R](authenticateOnUnauthorized: Boolean = true)
                                 (f: (String) => R): R = {
        // If there exists a current administrative token.
        var currentToken = adminToken
        // If there is no token or if the token has or is about to expire, renew
        // the token using the password credentials.
        if ((currentToken eq null) ||
            System.currentTimeMillis() + TokenExpirationGuard.toMillis +
            currentToken.serverIssueTime - currentToken.clientIssueTime >=
            currentToken.expirationTime) {

            if (isBlank(config.userName) || isBlank(config.password) ||
                isBlank(config.projectName)) {
                throw new KeystoneException(null,
                    "Keystone configuration must provide an administrative " +
                    "tenant, user name and password", null)
            }

            val auth = authenticate(config.projectName, config.userName,
                                    config.password)
            if (auth.token eq null) {
                throw new KeystoneUnauthorizedException(
                    null, "Keystone failed to return an authentication token", null)
            }

            currentToken = AdminToken(
                id = auth.tokenId,
                serverIssueTime = parseTimestamp(auth.token.issuedAt),
                clientIssueTime = System.currentTimeMillis(),
                expirationTime = parseExpiresAt(auth.token.expiresAt),
                isStatic = false)
            adminToken = currentToken
        }
        try return f(currentToken.id)
        catch {
            case _: KeystoneUnauthorizedException
                if authenticateOnUnauthorized && !adminToken.isStatic =>
                adminToken = null
        }
        withAdminToken(authenticateOnUnauthorized = false)(f)
    }

    /**
      * Handles the authentication response from the Keystone v3 API server and
      * returns and [[AuthResponse]] object.
      */
    private def handleAuthResponse(response: ClientResponse): AuthResponse = {
        val tokenId = response.getHeaders.getFirst("X-Subject-Token")

        if (response.getStatus < 300) {
            val token = response.getEntity(classOf[v3.KeystoneToken]).token
            return AuthResponse(tokenId = tokenId, r3 = token)
        }
        throw new KeystoneUnauthorizedException(response)
    }

}
